### PR TITLE
Logsumexp

### PR DESCRIPTION
**Pull request by unnonouno**
I implemented log-sum-exp function. That is used CRF loss function.
related to #1020 and #1092

### CODE REVIEW

**Comment by okuta** [comment]
Test condition is to tight.
https://ci.appveyor.com/project/pfnet/chainer/build/2244/job/4gqv7ijlwaq4un3e

**Comment by delta2323** [comment]
`logsumexp` is implemented in several places in `Chainer`. So I think we should replace them with this implementation, although we can do it as another PR.

**Comment by unnonouno** [comment]
@delta2323  I also think so

File: chainer/functions/math/logsumexp.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+        if self.axis is not None:
+            for axis in self.axis:
+                if axis >= 0:
+                    type_check.expect(
+                        axis < in_types[0].ndim,
+                    )
+                else:
+                    type_check.expect(
+                        -axis - 1 < in_types[0].ndim,
```
Comment (okuta) [inline chainer/functions/math/logsumexp.py]: How about `-axis <= in_types[0].ndim,`?

File: chainer/functions/math/logsumexp.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        return self.y,
+
+    def backward(self, inputs, grads):
+        xp = cuda.get_array_module(*inputs)
+        x, = inputs
+        gy, = grads
+
+        y = self.y
+        if self.axis is not None:
+            actual_axis = []
```
Comment (okuta) [inline chainer/functions/math/logsumexp.py]: How about the `actual_axis = [ax % x.ndim for ax in self.axis]`?

**Comment by okuta** [comment]
LGTM except comments.

File: chainer/functions/math/logsumexp.py (hunk 1 above)
Comment (unnonouno) [inline chainer/functions/math/logsumexp.py]: I copied it from `sum` function. I think consistency is more important.

File: chainer/functions/math/logsumexp.py (hunk 2 above)
Comment (unnonouno) [inline chainer/functions/math/logsumexp.py]: It is also from `sum`.

File: chainer/functions/math/logsumexp.py (hunk 1 above)
Comment (okuta) [inline chainer/functions/math/logsumexp.py]: OK, I understand.

File: chainer/functions/math/logsumexp.py (hunk 2 above)
Comment (okuta) [inline chainer/functions/math/logsumexp.py]: OK.

**Pull request merged**